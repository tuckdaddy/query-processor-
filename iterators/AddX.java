package iterators;


import java.util.Iterator;

public class AddX implements Iterator{
      
    // The Iterator that this apply  object will get its input from
    private final Iterator<Integer> input;	
    private final int x;
    // Constructs an instance of Addx with and int and an iterator
    public AddX(int x, Iterator<Integer> input) {
        this.input = input;
        this.x = x;
    }

    @Override
    public boolean hasNext() {
        return input.hasNext();
    }

    @Override
    // Method that overrides Iterators next() that adds the parameter x
    // to the data in input.next()
    public Integer next() {
        if (!hasNext()) throw new IllegalStateException();
        return input.next() + x;
    }
}       
