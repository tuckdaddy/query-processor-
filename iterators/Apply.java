package iterators;


import java.util.Iterator;

public class Apply<InT,OutT> implements Iterator<OutT> {
    // The function that will be applied to each input element to make an output element
    //Higher order function 
    // InT indicates input data, OutT indicates output data
    private final ApplyFunction<InT,OutT> f;

    // The Iterator that this Apply object will get its input from
    private final Iterator<InT> input;

    public Apply(ApplyFunction<InT, OutT> f, Iterator<InT> input) {
        this.f = f;
        this.input = input;
    }

    @Override
    public boolean hasNext() {
        return input.hasNext();
    }

    @Override
    public OutT next() {
        if((!hasNext())) throw new IllegalStateException();
        return f.apply(input.next());
    }
}
