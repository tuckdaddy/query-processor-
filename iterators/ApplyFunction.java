package iterators;

// Interface for classes to define a function from InT to OutT
public interface ApplyFunction<InT, OutT> {
		public OutT apply(InT x);
}
