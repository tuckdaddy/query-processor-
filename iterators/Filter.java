package iterators;

import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;
import java.lang.NullPointerException;

// Iterator that uses a Predicate to filter out elements from the input
public class Filter<T> extends FlatApply<T, T> {

    public Filter(Predicate<T> p, Iterator<T> input) {
        // you DO NOT need to modify the constructor
        super(new FilteringFlatApplyFunction<>(p), input);
    }

    // uses a Predicate to decide whether the input element is output or not
    private static class FilteringFlatApplyFunction<T> implements FlatApplyFunction<T, T> {
        private Predicate<T> derp; 
        public FilteringFlatApplyFunction(Predicate<T> p){
            this.derp = p;
        }
        
        // if x passes the predicate, add x to the new list 
        @Override 
        public List<T> apply(T x){
            List <T> lst = new LinkedList<>();
            boolean res = derp.check(x);
            if (res == true){
                lst.add(x);
                return lst;
            }
            else {
                return lst;
            }
        }
    }

    // You DO NOT need to define hasNext() and next(). FlatApply provides them already.
}