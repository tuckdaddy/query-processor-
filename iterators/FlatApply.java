package iterators;

import java.util.Queue;
import java.util.LinkedList;
import java.util.List;
import java.util.Iterator;

// Iterator that may produce 0 or more output elements for every input element
public class FlatApply<InT,OutT> implements Iterator<OutT> {
    private final Iterator <InT> input;
    private final FlatApplyFunction<InT,OutT> f;
    private final Queue<OutT> q;
    
    // Constructs a new linked list queue everytime constructor is called
    public FlatApply(FlatApplyFunction<InT,OutT> f, Iterator<InT> input) {		
        this.input = input;
        this.f = f;
        this.q = new LinkedList<>();
    }

    @Override
    // If the queue is not empty return true, else fill the queue by applying the 
    //  function until input.hasNext() is false
    //  but there is a check again if the queue is empty to eliminate any errors
    public boolean hasNext() {
        if (!q.isEmpty()){return true;}
        else{
            while(q.isEmpty() && input.hasNext()){
                List<OutT> res = f.apply(input.next());
                q.addAll(res);
            }
            if(q.isEmpty()) return false;
            return true;
        }
    }

    @Override
    public OutT next() {
        if((!hasNext())) throw new IllegalStateException();
        return q.poll();
    }
}
