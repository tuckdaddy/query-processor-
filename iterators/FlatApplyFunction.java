package iterators;


import java.util.List;

// Interface for classes to define a function from InT to List<OutT>
public interface FlatApplyFunction<InT, OutT> {
		public List<OutT> apply(InT x);
}
