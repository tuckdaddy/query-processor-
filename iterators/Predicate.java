package iterators;

// Interface for classes to define a function from T to boolean
public interface Predicate<T> {
	public boolean check(T data);		
}
