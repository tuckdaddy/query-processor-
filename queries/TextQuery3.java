package queries;


import iterators.Apply;
import iterators.ApplyFunction;
import iterators.Filter;
import iterators.FlatApply;
import iterators.FlatApplyFunction;
import iterators.Predicate;
import java.util.Arrays;
import java.util.Iterator;
import java.util.List;
import readers.TextFileReader;

// return all filenames that contain the word "Mars" 
public class TextQuery3 {
    public static void main(String[] args) {
        Iterator<Pair<String,String>> filenameAndContents = new TextFileReader("../sci.space");
        Iterator<String> fileName= new Apply(new TakeLeft<>(), filenameAndContents);
        Iterator<String> contents = new Apply(new TakeRight<>(), filenameAndContents);
        Iterator<String> words = new FlatApply<>(new SplitBy("[ .!,:<>@=*_ ()/\\t~#\"|-]"), contents); 
        Iterator<String> mars = new Filter<>(new FindMars("Mars"), words);
        /* finish the query using a combination of Applys and Filters */
	// Eventually the last iterator will go through the while loop and print statement
        //Finds files with words matching the key 'Mars'. Works correctly but doesnt match expected/TextQuery3.txt
        
        while (mars.hasNext()) {
            System.out.println(fileName.next());
            //To move the iterator to the next file containing 'Mars'
            mars.next();
        }
    }
    // put your classes that implement ApplyFunction and Predicate here
    private static class FindMars implements Predicate<String>{
        private String s = null;
        public FindMars(String s){
            this.s = s;
        }
        @Override
        public boolean check(String data){
            return s.equals(data);
        }
    }

    private static class TakeLeft<L,R> implements ApplyFunction<Pair<L,R>, L>{
        @Override
        public L apply(Pair<L,R> x){
            return x.left;
        }
    }
    private static class TakeRight<L,R> implements ApplyFunction<Pair<L,R>, R> {
        @Override
        public R apply(Pair<L, R> x) {
            return x.right;	
        }
    }
    public static class SplitBy implements FlatApplyFunction<String, String> {
		private String ch;
		public SplitBy(String c) {
			this.ch = c;
		}
		@Override
		public List<String> apply(String x) {
                    return Arrays.asList(x.split(ch));
		}
	}
}
