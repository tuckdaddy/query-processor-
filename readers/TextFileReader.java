package readers;


import queries.Pair;
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.util.Iterator;

public class TextFileReader implements Iterator<Pair<String,String>> {
		private final File[] files;
		private int curFile;

		public TextFileReader(String folderPath) {
			File folder = new File(folderPath);
			this.files = folder.listFiles();
			this.curFile = 0;
		}
		
		@Override
		public boolean hasNext() {
			return curFile < files.length;
		}

		/* returns a Pair containing (filename, contents), i.e.,
		   Pair.left=filename and Pair.right=contents
		*/
		@Override
		public Pair<String, String> next() {
			try {
				String name = files[curFile].getName();
				BufferedReader in = new BufferedReader(new FileReader(files[curFile]));
				StringBuilder sb = new StringBuilder();
				String s = in.readLine();
				while (s != null) {
					sb.append(s);
					s = in.readLine();
				}
				curFile++;
				return new Pair(name, sb.toString());
			} catch (IOException ex) {
				throw new RuntimeException(ex);
			}
		}
}
