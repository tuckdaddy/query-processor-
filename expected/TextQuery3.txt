59846
59848
59849
59870
59871
59873
59874
59905
59907
59908
59909
60103
60151
60170
60173
60187
60195
60210
60231
60832
60877
60922
60933
60960
60996
61004
61007
61012
61032
61040
61062
61110
61118
61129
61132
61141
61156
61178
61198
61209
61212
61221
61243
61253
61264
61316
61333
61383
61398
61401
61407
61409
61429
61432
61434
61435
61493
61516
61539
61546
61551
61567
61569
62116
62127
62374
62375
62405
